sound/soc/sof/ipc4-mtrace.c

```diff
@@ -0,0 +1,474 @@
+
+	priv->state_info.enable = 1;
+
+	msg.data_size = sizeof(priv->state_info);
+	msg.data_ptr = &priv->state_info;
+	if (iops->set_get_data(sdev, &msg, msg.data_size, true)) {
+		sdev->fw_trace_is_supported = false;
+		return 0;
+	}
+
+	priv->mtrace_is_enabled = true;
+
+	return 0;
+}
+
+static void ipc4_mtrace_disable(struct snd_sof_dev *sdev)
+{
+	struct sof_mtrace_priv *priv = sdev->fw_trace_data;
+	const struct sof_ipc_ops *iops = sdev->ipc->ops;
+	struct sof_ipc4_msg msg;
+	int i;
+
+	if (!priv->mtrace_is_enabled)
+		return;
+
+	msg.primary = SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG);
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MOD_ID(SOF_IPC4_MOD_INIT_BASEFW_MOD_ID);
+	msg.primary |= SOF_IPC4_MOD_INSTANCE(SOF_IPC4_MOD_INIT_BASEFW_INSTANCE_ID);
+	msg.extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_FW_PARAM_ENABLE_LOGS);
+
+	priv->state_info.enable = 0;
+
+	msg.data_size = sizeof(priv->state_info);
+	msg.data_ptr = &priv->state_info;
+	if (iops->set_get_data(sdev, &msg, msg.data_size, true))
+		sdev->fw_trace_is_supported = false;
+
+	priv->mtrace_is_enabled = false;
+
+	for (i = 0; i < sdev->num_cores; i++) {
+		struct sof_mtrace_core_data *core_data = &priv->cores[i];
+
+		core_data->host_read_ptr = 0;
+		core_data->dsp_write_ptr = 0;
+		wake_up(&core_data->trace_sleep);
+	}
+}
+
+/*
+ * Parse the slot descriptors at debug_box offset, we are only interested in the
+ * type of the slot to map it to a core
+ */
+static void sof_mtrace_find_core_slots(struct snd_sof_dev *sdev)
+{
+	struct sof_mtrace_priv *priv = sdev->fw_trace_data;
+	struct sof_mtrace_core_data *core_data;
+	u32 slot_desc_offset, type, core;
+	int i;
+
+	slot_desc_offset = sdev->debug_box.offset;
+	for (i = 0; i < MAX_MTRACE_SLOTS - 1; i++) {
+		/* Read the slot type */
+		sof_mailbox_read(sdev, slot_desc_offset + 4, &type, 4);
+		if ((type & SOF_MTRACE_SLOT_TYPE_MASK) == SOF_MTRACE_SLOT_DEBUG_LOG) {
+			core = type & SOF_MTRACE_SLOT_CORE_MASK;
+
+			if (core >= sdev->num_cores)
+				continue;
+
+			core_data = &priv->cores[core];
+			core_data->slot_offset = sdev->debug_box.offset;
+			core_data->slot_offset += SOF_MTRACE_SLOT_SIZE * (i + 1);
+			dev_dbg(sdev->dev, "%s: slot%d is used for core%u\n",
+				__func__, i, core);
+			if (core_data->missed_update) {
+				sof_ipc4_mtrace_update_pos(sdev, core);
+				core_data->missed_update = false;
+			}
+		} else if (type) {
+			dev_dbg(sdev->dev, "%s: slot%d is not a log slot (%#x)\n",
+				__func__, i, type);
+		}
+
+		slot_desc_offset += sizeof(u32) * 3;
+	}
+}
+
+static int ipc4_mtrace_init(struct snd_sof_dev *sdev)
+{
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct sof_mtrace_priv *priv;
+	int i, ret;
+
+	if (sdev->fw_trace_data) {
+		dev_err(sdev->dev, "fw_trace_data has been already allocated\n");
+		return -EBUSY;
+	}
+
+	if (!ipc4_data->mtrace_log_bytes ||
+	    ipc4_data->mtrace_type != SOF_IPC4_MTRACE_INTEL_TGL) {
```

Comment:
Is mtrace_log_bytes a fixed number? what is the use of it's value? thanks,

Reply:
It is supposed to be the size of the debug window, but on TGL it is coming back as the size of _one_ slot in the window, on APL it is 0x2000 (2 cores, 2 slots, I think).
If it is 0 then we conclude that the firmware does not supports mtrace type of logging (this is the case atm with SOF.IPC4).
On TGL type of platforms we _know_ that the trace window have 16 slots, it is static, so if the number is not NULL, mtrace can be enabled.